### PR TITLE
Implement ProtoMessage encoding for ConnectionEnd

### PR DESCRIPTION
**Pull request by ljoss17**
This PR implements the following:

- [ ] `ProtoMessage` for `ConnectionEnd`
- [ ] `ProtoMessage` for `Channel`

### CODE REVIEW

**Review by rnbguy — APPROVED** [review]
I left some comments. Feel free to ignore the ones about gas optimization.

File: cairo-libs/packages/protobuf/src/varint.cairo [hunk 1]
```diff
@@ -35,6 +35,23 @@ pub fn encode_varint_to_byte_array(value: u64) -> ByteArray {
     result
 }
 
+#[inline]
+pub fn encode_varintu128_to_byte_array(value: u128) -> ByteArray {
```
[the remark this answers is not in the document]
Comment (ljoss17) [reply cairo-libs/packages/protobuf/src/varint.cairo:39]: Same question here

File: cairo-libs/packages/protobuf/src/types/message.cairo [hunk 2]
```diff
@@ -54,19 +54,17 @@ pub impl EncodeContextImpl of EncodeContextTrait {
     fn encode_repeated_field<T, +ProtoMessage<T>>(
         ref self: EncodeContext, field_number: u8, value: @Array<T>,
     ) {
-        let mut i = 0;
-        while i < value.len() {
+        let mut value_span = value.span();
+        while let Option::Some(v) = value_span.pop_front() {
```
[the remark this answers is not in the document]
Comment (ljoss17) [reply cairo-libs/packages/protobuf/src/types/message.cairo:58]: This seems to be more costly gas-wise

File: cairo-libs/packages/protobuf/src/primitives/numeric.cairo [hunk 3]
```diff
@@ -7,7 +7,27 @@ use protobuf::types::message::{
     ProtoMessage, decode_raw,
 };
 use protobuf::types::tag::WireType;
-use protobuf::varint::{decode_varint_from_byte_array, encode_varint_to_byte_array};
+use protobuf::varint::{
+    decode_varint_from_byte_array, decode_varintu128_from_byte_array, encode_varint_to_byte_array,
+    encode_varintu128_to_byte_array,
+};
+
+pub impl U128AsProtoMessage of ProtoMessage<u128> {
```
[the remark this answers is not in the document]
Comment (ljoss17) [reply cairo-libs/packages/protobuf/src/primitives/numeric.cairo:16]: Done b1b1937779b525a19c4b12799b1f05d92fdcee6a

File: cairo-libs/packages/protobuf/src/primitives/array.cairo [hunk 4]
```diff
@@ -92,13 +79,12 @@ pub impl ArrayAsProtoMessage<T, +ProtoMessage<T>, +Drop<T>, +Default<T>> of Prot
...
-        let mut i = 0;
         if self.len() == 0 {
             context.buffer.append_byte(0);
         }
-        while i < self.len() {
-            context.buffer.append_byte(self[i].clone());
-            i += 1;
+        let mut self_span = self.span();
+        while let Option::Some(item) = self_span.pop_front() {
+            context.buffer.append_byte(*item);
```
[the remark this answers is not in the document]
Comment (ljoss17) [reply cairo-libs/packages/protobuf/src/primitives/array.cairo:87]: Same as https://github.com/informalsystems/ibc-starknet/pull/381#discussion_r2044702208

File: cairo-libs/packages/protobuf/src/types/message.cairo (hunk 2 above)
[the remark this answers is not in the document]
Comment (rnbguy) [reply cairo-libs/packages/protobuf/src/types/message.cairo:58]: ah good to know 👍🏼

File: cairo-libs/packages/protobuf/src/varint.cairo [hunk 5]
```diff
@@ -89,3 +105,26 @@ pub fn decode_varint_from_byte_array(bytes: @ByteArray, ref index: usize) -> Res
     Result::Ok(value)
 }
 
+#[inline]
+pub fn decode_varintu128_from_byte_array(
```
[the remark this answers is not in the document]
Comment (rnbguy) [reply cairo-libs/packages/protobuf/src/varint.cairo:109]: yes. do you see any problem? (I may not have all the context)

File: cairo-libs/packages/protobuf/src/primitives/array.cairo [hunk 6]
```diff
@@ -27,61 +27,48 @@ pub impl ByteArrayAsProtoMessage of ProtoMessage<ByteArray> {
...
-                i += 1;
-            };
-        } else {
-            while i < self.len() {
-                self[i].encode_raw(ref context);
-                i += 1;
-            };
+            } else {
+                item.encode_raw(ref context);
+            }
```
[the remark this answers is not in the document]
Comment (ljoss17) [reply cairo-libs/packages/protobuf/src/primitives/array.cairo]: Done in 72edcadfbd5fcec8da9ebf4380c5c46f1ea94fbc

File: cairo-libs/packages/protobuf/src/primitives/array.cairo [hunk 7]
```diff
@@ -92,13 +79,12 @@ pub impl ArrayAsProtoMessage<T, +ProtoMessage<T>, +Drop<T>, +Default<T>> of Prot
 
 pub impl BytesAsProtoMessage of ProtoMessage<Array<u8>> {
     fn encode_raw(self: @Array<u8>, ref context: EncodeContext) {
-        let mut i = 0;
         if self.len() == 0 {
             context.buffer.append_byte(0);
```
[the remark this answers is not in the document]
Comment (rnbguy) [reply cairo-libs/packages/protobuf/src/primitives/array.cairo:85]: we need to confirm if this breaks anything. can you possibly add a test for this with an empty `bytes` and check it with prost?

File: cairo-libs/packages/ics23/src/verify.cairo [hunk 8]
```diff
@@ -177,35 +175,45 @@ fn ensure_inner(inner: @InnerOp, spec: ProofSpec) {
...
-    while i < spec_prefix.len() {
-        expected.append(*proof_prefix[i]);
-        i += 1;
+
+    let mut proof_prefix_span = proof_prefix.span();
+    let mut spec_prefix_span = spec_prefix.span();
+
+    let mut has_prefix = true;
+
+    while let Option::Some(s_prefix) = spec_prefix_span.pop_front() {
```
[the remark this answers is not in the document]
Comment (rnbguy) [reply cairo-libs/packages/ics23/src/verify.cairo:190]: marking as resolved as probably higher gas use.

File: cairo-libs/packages/cometbft/src/utils.cairo [hunk 9]
```diff
@@ -114,11 +114,18 @@ pub fn u32_8_to_array_u8(u32_8: [u32; 8]) -> Array<u8> {
...
+    while i != 32 {
+        if power >= num {
+            break;
+        }
+        power *= 2;
+        i += 1;
     }
-    two_power
+
+    power
```
[the remark this answers is not in the document]
Comment (rnbguy) [reply cairo-libs/packages/cometbft/src/utils.cairo:128]: https://github.com/informalsystems/ibc-starknet/pull/387/files#diff-de1fe0d7f8fb2cd87190d55fbc143cf255a06decbbf434649220ad0051905febR117-R131

File: cairo-libs/packages/protobuf/src/varint.cairo (hunk 5 above)
[the remark this answers is not in the document]
Comment (ljoss17) [reply cairo-libs/packages/protobuf/src/varint.cairo:109]: encode and decode u64 is used for `U64AsProtoMessage`, https://github.com/informalsystems/ibc-starknet/blob/b1b1937779b525a19c4b12799b1f05d92fdcee6a/cairo-libs/packages/protobuf/src/primitives/numeric.cairo#L37

File: cairo-libs/packages/protobuf/src/hex.cairo [hunk 10]
```diff
@@ -34,10 +35,13 @@ pub fn encode(input: @ByteArray) -> ByteArray {
...
 pub fn decode(input: @ByteArray) -> ByteArray {
-    assert(input.len() % 2 == 0, 'Invalid hex string length');
+    let input_len = input.len();
+    assert(input_len % 2 == 0, 'Invalid hex string length');
     let mut output = "";
     let mut i = 0;
-    while i < input.len() {
+    // Since input_len % 2 == 0, we know i += 2 will eventually be
+    // equal to input_len
+    while i != input_len {
```
[the remark this answers is not in the document]
Comment (ljoss17) [reply cairo-libs/packages/protobuf/src/hex.cairo]: Done in c1a0b25bc85cd9a5f544dbfae3939f52a7692048

File: cairo-libs/packages/protobuf/src/base64.cairo [hunk 11]
```diff
@@ -83,7 +83,9 @@ pub fn decode(input: @ByteArray) -> ByteArray {
     let mut output = "";
     let mut i = 0;
 
-    while i < input_len {
+    // Since input_len % 4 == 0, we know i += 4 will eventually be
+    // equal to input_len
+    while i != input_len {
```
[the remark this answers is not in the document]
Comment (ljoss17) [reply cairo-libs/packages/protobuf/src/base64.cairo]: Done in c1a0b25bc85cd9a5f544dbfae3939f52a7692048

File: cairo-libs/packages/ics23/src/verify.cairo [hunk 12]
```diff
@@ -287,34 +296,34 @@ fn order_from_padding(inner_spec: InnerSpec, inner_op: @InnerOp) -> u32 {
...
+                let from = actual_prefix.unwrap() + child_size * o.clone();
                 let mut expected_prefix = ArrayTrait::new();
                 for i in from..from + child_size {
                     expected_prefix.append(*inner_op.prefix[i]);
                 }
-                if inner_spec.empty_child != expected_prefix {
-                    are_empty = false;
-                    break;
+                if inner_spec_empty_child != expected_prefix {
+                    return false;
```
[the remark this answers is not in the document]
Comment (ljoss17) [reply cairo-libs/packages/ics23/src/verify.cairo:321]: No errors when compiling

File: cairo-libs/packages/ics23/src/utils.cairo [hunk 13]
```diff
@@ -246,13 +248,15 @@ pub enum Ordering {
...
+    let mut lhs_span = lhs.span();
+    let mut rhs_span = rhs.span();
     let min_len = core::cmp::min(lhs_len, rhs_len);
 
     let mut ordering = Ordering::Equal;
 
-    for i in 0..min_len {
-        let l = lhs.at(i);
-        let r = rhs.at(i);
+    for _ in 0..min_len {
```
[the remark this answers is not in the document]
Comment (ljoss17) [reply cairo-libs/packages/ics23/src/utils.cairo]: Done in 776e5b684fc92d3adc1b2f359b235363133c64bd

File: cairo-libs/packages/cometbft/src/utils.cairo (hunk 9 above)
[the remark this answers is not in the document]
Comment (ljoss17) [reply cairo-libs/packages/cometbft/src/utils.cairo]: Done in 0588e7e667d102ac7ed274f1414977fb433e355f

File: cairo-contracts/packages/core/src/host/identifiers.cairo [hunk 14]
```diff
@@ -36,7 +41,7 @@ pub impl ClientIdZero of Zero<ClientId> {
     }
 }
 
-#[derive(Clone, Debug, Drop, PartialEq, Serde, starknet::Store)]
+#[derive(Default, Clone, Debug, Drop, PartialEq, Serde, starknet::Store)]
```
[the remark this answers is not in the document]
Comment (rnbguy) [reply cairo-contracts/packages/core/src/host/identifiers.cairo:44]: aah I see.

File: cairo-libs/packages/protobuf/src/primitives/array.cairo (hunk 7 above)
[the remark this answers is not in the document]
Comment (ljoss17) [reply cairo-libs/packages/protobuf/src/primitives/array.cairo:85]: Opened https://github.com/informalsystems/ibc-starknet/issues/401
